File: website/docs-examples/3d/c/examples/common_recipes3.c
#include "snippets.h"
#include <math.h>

// DOCUSAURUS: OneWayPlatform start
typedef struct OneWayPlatform {
    R3ColliderHandle platform;
} OneWayPlatform;

static int same_collider(R3ColliderHandle a, R3ColliderHandle b) {
    return a.world == b.world && a.index == b.index && a.generation == b.generation;
}

static void RAPIER_CALL one_way_platform(void *user_data, const R3ReadContext *read,
                                         R3ColliderHandle collider1, R3ColliderHandle collider2,
                                         R3ContactModificationContext *context) {
    (void)read;
    (void)collider2;
    const OneWayPlatform *hook = user_data;
    /* Keep only the contacts pushing along the local +y axis of the platform; the other
     * ones (the character arriving from below) are discarded. The normal is expressed in
     * the frame of the first collider of the pair, hence the flip. */
    R3Vector allowed_local_n1 =
        same_collider(collider1, hook->platform) ? r3Vector(0.0, 1.0, 0.0) : r3Vector(0.0, -1.0, 0.0);
    r3ContactModificationContext_UpdateAsOnewayPlatform(context, allowed_local_n1, 0.1);
}
// DOCUSAURUS: OneWayPlatform stop

// DOCUSAURUS: ConveyorBelt start
static void RAPIER_CALL conveyor_belt(void *user_data, const R3ReadContext *read,
                                      R3ColliderHandle collider1, R3ColliderHandle collider2,
                                      R3ContactModificationContext *context) {
    (void)user_data;
    (void)read;
    (void)collider1;
    (void)collider2;
    /* The belt drags the objects along the world-space z axis at 12 m/s. */
    r3ContactModificationContext_SetTangentVelocity(context, r3Vector(0.0, 0.0, 12.0));
}
// DOCUSAURUS: ConveyorBelt stop

int main(void) {
    snippets_init();
    R3World *world = r3NewWorld();

    /* A box falling on the platform, so the hooks have contacts to modify. */
    R3RigidBodyDesc box_body = r3DynamicRigidBodyDesc();
    box_body.position.translation = r3Vector(0.0, 1.5, 0.0);
    R3ColliderDesc box = r3CuboidColliderDesc(r3Vector(0.25, 0.25, 0.25));
    r3InsertCollider(r3InsertRigidBody(world, &box_body), &box);

    // DOCUSAURUS: MovingPlatform start
    R3RigidBodyDesc platform_body = r3KinematicPositionBasedRigidBodyDesc();
    platform_body.position.translation = r3Vector(0.0, 1.0, 0.0);
    R3RigidBodyHandle platform_handle = r3InsertRigidBody(world, &platform_body);
    R3ColliderDesc platform_collider_desc = r3CuboidColliderDesc(r3Vector(2.0, 0.1, 2.0));
    r3InsertCollider(platform_handle, &platform_collider_desc);

    for (int step = 0; step < 200; step++) {
        /* Setting the next position of the platform, once per timestep. */
        R3Real time = (R3Real)step * r3TimeStep(world);
        r3RigidBody_SetNextKinematicTranslation(platform_handle, r3Vector(sinf(time) * 2.0f, 1.0, 0.0));
        r3Step(world, NULL, NULL);
    }
    // DOCUSAURUS: MovingPlatform stop

    // DOCUSAURUS: Hooks start
    /* The hooks are only called for the colliders asking for them. */
    R3ColliderHandle platform_collider;
    r3RigidBody_Colliders(platform_handle, &platform_collider, 1);
    r3Collider_SetActiveHooks(platform_collider, R3_MODIFY_SOLVER_CONTACTS);

    OneWayPlatform platform = {platform_collider};
    R3PhysicsHooks hooks = {0};
    hooks.user_data = &platform;
    hooks.modify_solver_contacts_context = one_way_platform;
    r3Step(world, &hooks, NULL);
    // DOCUSAURUS: Hooks stop

    R3PhysicsHooks belt_hooks = {0};
    belt_hooks.modify_solver_contacts_context = conveyor_belt;
    r3Step(world, &belt_hooks, NULL);

    r3FreeWorld(world);
    return EXIT_SUCCESS;
}

File: website/docs-examples/3d/bevy/examples/common_recipes3.rs
use bevy::ecs::system::SystemParam;
use bevy::prelude::*;
use bevy_rapier3d::prelude::*;

// DOCUSAURUS: OneWayPlatform start
#[derive(Component)]
struct OneWayPlatform;

#[derive(SystemParam)]
struct OneWayPlatformHooks<'w, 's> {
    platforms: Query<'w, 's, (), With<OneWayPlatform>>,
}

impl BevyPhysicsHooks for OneWayPlatformHooks<'_, '_> {
    fn modify_solver_contacts(&self, mut context: ContactModificationContextView) {
        // Keep only the contacts pushing along the local +y axis of the platform; the other
        // ones (the character arriving from below) are discarded. The normal is expressed in
        // the frame of the first collider of the pair, hence the flip.
        let allowed_local_n1 = if self.platforms.contains(context.collider1()) {
            Vec3::Y
        } else if self.platforms.contains(context.collider2()) {
            -Vec3::Y
        } else {
            return;
        };
        context.update_as_oneway_platform(allowed_local_n1, 0.1);
    }
}
// DOCUSAURUS: OneWayPlatform stop

#[allow(dead_code)]
// DOCUSAURUS: ConveyorBelt start
#[derive(SystemParam)]
struct ConveyorBeltHooks<'w, 's> {
    belts: Query<'w, 's, (), With<ConveyorBelt>>,
}

#[derive(Component)]
struct ConveyorBelt;

impl BevyPhysicsHooks for ConveyorBeltHooks<'_, '_> {
    fn modify_solver_contacts(&self, mut context: ContactModificationContextView) {
        if !self.belts.contains(context.collider1()) && !self.belts.contains(context.collider2())
        {
            return;
        }
        if let Some(contacts) = context.solver_contacts_mut() {
            for contact in contacts.iter_mut() {
                // The belt drags the objects along the world-space z axis at 12 m/s.
                contact.tangent_velocity.z = 12.0;
            }
        }
    }
}
// DOCUSAURUS: ConveyorBelt stop

#[derive(Component)]
struct MovingPlatform;

fn main() {
    App::new()
        // DOCUSAURUS: HooksPlugin start
        .add_plugins((
            DefaultPlugins,
            // The hooks are given to the physics plugin.
            RapierPhysicsPlugin::<OneWayPlatformHooks>::default(),
        ))
        // DOCUSAURUS: HooksPlugin stop
        .add_systems(
            Startup,
            (
                setup_one_way_platform,
                setup_moving_platform,
                setup_conveyor_belt,
            ),
        )
        .add_systems(Update, move_platform)
        .run();
}

// DOCUSAURUS: Hooks start
fn setup_one_way_platform(mut commands: Commands) {
    // The hooks are only called for the colliders asking for them.
    commands.spawn((
        Transform::from_xyz(0.0, 1.0, 0.0),
        Collider::cuboid(2.0, 0.1, 2.0),
        ActiveHooks::MODIFY_SOLVER_CONTACTS,
        OneWayPlatform,
    ));
}
// DOCUSAURUS: Hooks stop

fn setup_conveyor_belt(mut commands: Commands) {
    commands.spawn((
        Transform::from_xyz(0.0, 0.0, -5.0),
        Collider::cuboid(2.0, 0.1, 2.0),
        ActiveHooks::MODIFY_SOLVER_CONTACTS,
        ConveyorBelt,
    ));
}

// DOCUSAURUS: MovingPlatform start
fn setup_moving_platform(mut commands: Commands) {
    commands.spawn((
        RigidBody::KinematicPositionBased,
        Transform::from_xyz(0.0, 1.0, 5.0),
        Collider::cuboid(2.0, 0.1, 2.0),
        MovingPlatform,
    ));
}

fn move_platform(time: Res<Time>, mut platforms: Query<&mut Transform, With<MovingPlatform>>) {
    for mut transform in platforms.iter_mut() {
        // Setting the next position of the platform: the physics engine derives the
        // velocity needed to reach it at the end of the next timestep.
        transform.translation.x = time.elapsed_secs().sin() * 2.0;
    }
}
// DOCUSAURUS: MovingPlatform stop
